### ISSUE TITLE
Decimal::to_f64 produces imprecise results

### ISSUE THREAD

**Issue by nappa85** [paupino/rust-decimal#401]
I'm having some problems with your implementation of to_f64,

for example:
```rust
assert_eq!(Decimal::new(22238, 4).to_f64(), Decimal::new(22238, 4).to_string().parse::<f64>().ok());
```

fails, your implementation returns 2.2237999999999998 while std returns correctly 2.2238

Is there a way to make your implementation more precise?

**Comment by schungx** [comment]
You're really asking for a strict adherence to IEEE conversion rules for `Decimal`.  Not sure if this is intended for this crate.

In actual usage, you _never_ want to depend on the _exact_ float representation.  If you depend on decimal point to the 13th place, there is something wrong, since your number only has 4 significant digits.

I am quite sure that, if you need such, you're probably not using floating-points correctly.

**Comment by nappa85** [comment]
My problem is on serialization, I know that, to not lose precision, the default is to serialize as string.
For backward compatibility with a service I'm replacing, I have to serialize as float.
Using the feature "serde_float", that internally uses to_f64(), gives me those problems.
Instead, using Decimal::as_string() and then parsing with std gives me the correct result.
That been said, it isn't a problem of value not rapresentable with f64, it's a conversion problem of Decimal::to_f64() implementation.
I can accept the reply "to_f64() is a best-effort implementation and we can't do better that this"

**Comment by schungx** [comment]
But my question is...  when you work with floats, you automatically know that you're working with an imprecise representation.  f64 is imprecise, therefore you cannot just test them for equality.  You need to provide a tolerance.

So what I'm saying is, `2.2237999999999998` and `2.2238` are exactly the _same_ number (given a required precision).  So it doesn't matter which form it is serialized to, because they are to represent essentially the same number.  There is no "correct" result; both are "correct".  The two representations are equivalent up to your required precision.

You can only say that different implementations of decimal conversions have slightly different behavior, and they give different strings, but all should fall within the required precision.

**Comment by schungx** [comment]
Just FYI... `2.22379999999999977688958097133` and `2.22380000000000022097879082139` differ by one single bit - the least significant one.  The last bit.

Therefore, 2.2238 falls exactly between these two representations.  Now the second one is "closer"... but it doesn't mean the first one is incorrect.  It is correct if you're working with 5 significant digits, or up to 16 (who needs 15 decimal points?) You're talking about the last bit in a 64-bit floating point number.

**Comment by nappa85** [comment]
So you're suggesting to do something similar, e.g. let's say I want 4 decimals:
 ```rust
Decimal::to_f64().map(|f| (f * 10000.0).round() / 10000.0)
 ```

**Comment by schungx** [comment]
Not really.  You can just do what you originally did.  What I mean is that, of the two `f64` values you got from the two methods, although they are different in bit patterns, they are _the same_ floating-point number for four decimal places (in your case).

In fact, you need to think in terms of "significant digits" instead of decimal places.  That's how the IEEE floating-point format is defined upon.

**Comment by c410-f3r** [comment]
> Just FYI... `2.22379999999999977688958097133` and `2.22380000000000022097879082139` differ by one single bit - the least significant one. The last bit.

This is gold. Thanks!

**Comment by paupino** [comment]
While this conversation makes sense, I _do_ think there is a small bug in the `to_f64` implementation whereby the final step which intends to "round" the number to the closest viable representation isn't actually doing what is being intended. Since there are many eyes on this I'll push up the proposed change and get some eyes across it.

**Comment by paupino** [comment]
Fixed in #408
